### PR TITLE
Dns discovery custom resolver

### PR DESCRIPTION
**Pull request by chbatey**
Includes importing service discovery.

Refs https://github.com/akka/akka/issues/25931

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -474,6 +474,14 @@ lazy val actorTypedTests = akkaModule("akka-actor-typed-tests")
...
   .enablePlugins(NoPublish)
 
+lazy val discovery = akkaModule("akka-discovery")
+  .dependsOn(
+    actor,
+    testkit % "test->test"
+  )
+  .settings(Dependencies.discovery)
+  .settings(AkkaBuild.mayChangeSettings)
+  .settings(AutomaticModuleName.settings("akka.discovery"))
```
Comment (patriknw) [inline build.sbt]: OSGi is missing

File: akka-docs/src/main/paradox/discovery/index.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
+# Discovery
+
+@@@ warning
+
+This module is currently marked as @ref:[may change](../common/may-change.md)
+This means that API or semantics can
+change without warning or deprecation period and it is not recommended to use
+this module in production.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/discovery/index.md]: Can't we remove that part of not for production? We know this is and will be heavily used in production. A module can be marked may change but still be ready for for production.

File: akka-docs/src/main/paradox/discovery/index.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+* `serviceName`, `portName` and `protocol` set: SRV query in the form: `_port._protocol._name` Where the `_`s are added.
+* Any query  missing any of the fields is mapped to a A/AAAA query for the `serviceName`
+
+The mapping between Akka service discovery terminology and SRV terminology:
+
+* SRV service = port
+* SRV name = serviceName
+* SRV protocol = protocol
+
+Configure it to be used as discovery implementation in your `application.conf` and `async-dns` to be uses
```
Comment (patriknw) [inline akka-docs/src/main/paradox/discovery/index.md]: "to be uses" ?

File: akka-docs/src/main/paradox/discovery/index.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+@@@
+
+Akka Discovery provides an interface around various ways of locating services, such as DNS
+or using configuration or key-value stores like zookeeper, consul. The built in methods are:
+
+* Configuration
+* DNS
+* Aggregate
+
+In addition `akka-management` contains methods for:
```
Comment (patriknw) [inline akka-docs/src/main/paradox/discovery/index.md]: link to Akka Management documentation

File: akka-docs/src/main/paradox/discovery/index.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+* Kubernetes API
+* AWS
+* Consul
+* Marathon API
+
+Discovery used to be part of `akka-management` but has become an Akka module as of 2.5.19 of Akka and version 0.20
+of `akka-management`.
```
Comment (patriknw) [inline akka-docs/src/main/paradox/discovery/index.md]: Akka Management vs `akka-management`?

File: akka-docs/src/main/paradox/discovery/index.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+* Any query  missing any of the fields is mapped to a A/AAAA query for the `serviceName`
+
+The mapping between Akka service discovery terminology and SRV terminology:
+
+* SRV service = port
+* SRV name = serviceName
+* SRV protocol = protocol
+
+Configure it to be used as discovery implementation in your `application.conf` and `async-dns` to be uses
+as the Akka DNS resolver:
```
Comment (patriknw) [inline akka-docs/src/main/paradox/discovery/index.md]: changing to async-dns had other possibly undesired consequences and now it loads async dns anyway, so should it be removed from this docs, or come with warnings?

Comment (chbatey) [reply akka-docs/src/main/paradox/discovery/index.md]: Removed the note to make it the akka resolver

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    system.settings.config.getString("akka.discovery.method") match {
+      case "<method>" ⇒
+        throw new IllegalArgumentException(
+          "No default service discovery implementation configured in " +
+            "`akka.discovery.method`. Make sure to configure this setting to your preferred implementation such as " +
+            "'akka-dns' in your application.conf (from the akka-discovery-dns module).")
+      case method ⇒ method
+    }
+
+  private lazy val _simpleImpl = loadServiceDiscovery(_defaultImplMethod)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: Doh

File: akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * A service lookup. It is up to each mechanism to decide
+ * what to do with the optional portName and protocol fields.
+ * For example `portName` could be used to distinguish between
+ * Akka remoting ports and HTTP ports.
+ *
+ */
+@ApiMayChange
+case class Lookup(serviceName: String, portName: Option[String], protocol: Option[String]) {
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/ServiceDiscovery.scala]: Doh

File: akka-discovery/src/main/scala/akka/discovery/aggregate/AggregateServiceDiscovery.scala [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import akka.event.Logging
+import akka.util.Helpers.Requiring
+import com.typesafe.config.Config
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+final class AggregateSimpleServiceDiscoverySettings(config: Config) {
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/aggregate/AggregateServiceDiscovery.scala]: Pesky comments and there was also a "simple" lookup which i got rid of

File: build.sbt (hunk 1 above)
Comment (chbatey) [reply build.sbt]: i'll pretend I know what that does :)

File: akka-discovery/src/main/resources/reference.conf [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    # ["akka-dns" , "akka-confg" ]
+    discovery-mechanisms = []
+
+  }
+
+  # DNS based service discovery
+  akka-dns {
+    class = akka.discovery.dns.DnsServiceDiscovery
+  }
+}
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-discovery/src/main/resources/reference.conf]: What's the best way to do that? We can't rely on it still being a jar as someone may have created a fat jar, should we try load a class from there that's not in here?

File: akka-docs/src/main/paradox/discovery/index.md (hunk 5 above)
Comment (chbatey) [reply akka-docs/src/main/paradox/discovery/index.md]: Will switch to Akka Management

File: akka-docs/src/main/paradox/index.md [hunk 11]
```diff
@@ -18,5 +18,6 @@
 * [howto](howto.md)
 * [project/index](project/index.md)
 * [additional/index](additional/index.md)
+* [discovery](discovery/index.md)
```
[the remark this answers is not in the document]
Comment (chbatey) [reply akka-docs/src/main/paradox/index.md]: Not sure, most of the things in networking are quite low level. What do others think?

File: akka-discovery/src/main/resources/reference.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-discovery/src/main/resources/reference.conf]: Is there some reference.conf property or class that only exists in the old version? Then we can look for that and fail.

[the remark this answers is not in the document]
Comment (chbatey) [reply akka-discovery/src/main/resources/reference.conf]: Plenty of classes as we re-named from from simple service discovery.

File: akka-actor/src/main/scala/akka/io/Dns.scala [hunk 12]
```diff
@@ -87,32 +89,77 @@ object Dns extends ExtensionId[DnsExt] with ExtensionIdProvider {
...
+   * be used in this case
+   */
+  @InternalApi
+  private[akka] def loadAsyncDns(managerName: String): ActorRef = {
+    // This can't pass in `this` as then AsyncDns would pick up the system settings
+    asyncDns.computeIfAbsent(managerName, new JFunction[String, ActorRef] {
+      override def apply(r: String): ActorRef = {
+        val settings = new Settings(system.settings.config.getConfig("akka.io.dns"), "async-dns")
+        val provider = system.dynamicAccess.getClassFor[DnsProvider](settings.ProviderObjectName).get.newInstance()
+        system.log.info("Creating async dns resolver {} with manger name {}", settings.Resolver, managerName)
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/io/Dns.scala]: 'manager'

File: akka-discovery/src/main/resources/reference.conf [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+akka.discovery {
+
+  # Users MUST configure this value to set the default discovery method.
+  #
+  # The value can be an implementation config path name, such as "akka-dns",
+  # which would attempt to resolve as `akka.discovery.akka-dns` which is expected
+  # to contain a `class` setting. As fallback, the root `akka-dns` setting scope
+  # would be used. If none of those contained a `class` setting, then the value is
+  # assumed to be a class name, and an attempt is made to intanciate it.
```
Comment (raboof) [inline akka-discovery/src/main/resources/reference.conf]: 'instantiate'

File: akka-docs/src/main/paradox/discovery/index.md [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+
+@@@
+
+Akka Discovery provides an interface around various ways of locating services. The built in methods are:
+
+* Configuration
+* DNS
+* Aggregate
+
+In addition [Akka Management](https://developer.lightbend.com/docs/akka-management/current/) contains methods for:
```
Comment (raboof) [inline akka-docs/src/main/paradox/discovery/index.md]: Perhaps 'such as', so we can add more mechanisms there without having to update the list here?

File: akka-docs/src/main/paradox/discovery/index.md [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+
+* Kubernetes API
+* AWS
+* Consul
+* Marathon API
+
+Discovery used to be part of Akka Management but has become an Akka module as of `2.5.19` of Akka and version `0.21.0`
+of Akka Management. If you're also using Akka Management for other service discovery methods or bootstrap make
+sure you are using at least version
```
Comment (raboof) [inline akka-docs/src/main/paradox/discovery/index.md]: `0.21.0` I guess?

File: akka-docs/src/main/paradox/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (raboof) [reply akka-docs/src/main/paradox/index.md]: Having it as a top-level section now might be OK, we can always group it somewhere else later.

**Review by raboof — COMMENTED** [review]
Looks good, I agree some way to proactively detect and warn that an incompatible version of the discovery from akka-management is used would be useful.

File: akka-discovery/src/main/resources/reference.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (TimMoore) [reply akka-discovery/src/main/resources/reference.conf]: Great idea!

File: akka-docs/src/main/paradox/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply akka-docs/src/main/paradox/index.md]: feels misplaced at the end, at least move it up after utilities

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+import akka.discovery.ServiceDiscovery.{ Resolved, ResolvedTarget }
+import akka.event.Logging
+import com.typesafe.config.Config
+
+import scala.collection.JavaConverters._
+import scala.collection.{ breakOut, immutable }
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+private object ConfigServicesParser {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: might still be public for Java? then add full InternalApi markers

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        (serviceName, Resolved(serviceName, resolvedTargets))
+    }
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+class ConfigServiceDiscovery(system: ExtendedActorSystem) extends ServiceDiscovery {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: can also have `private[akka]`

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+  private val log = Logging(system, getClass)
+
+  private val resolvedServices = ConfigServicesParser.parse(
+    system.settings.config.getConfig(system.settings.config.getString("akka.discovery.config.services-path"))
+  )
+
+  log.debug("Config discovery serving: {}", resolvedServices)
+
+  override def lookup(lookup: Lookup, resolveTimeout: FiniteDuration): Future[Resolved] = {
+    Future.successful(resolvedServices.getOrElse(lookup.serviceName, Resolved(lookup.serviceName, immutable.Seq.empty)))
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: `immutable.Seq.empty` a.k.a. `Nil`

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+      .toMap
+
+    byService.map {
+      case (serviceName, full) ⇒
+        val endpoints = full.getConfigList("endpoints").asScala
+        val resolvedTargets: immutable.Seq[ResolvedTarget] = endpoints.map { c ⇒
+          val host = c.getString("host")
+          val port = if (c.hasPath("port")) Some(c.getInt("port")) else None
+          ResolvedTarget(host = host, port = port, address = None)
+        }(breakOut)
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: perhaps not introduce more `breakOut`

File: akka-discovery/src/main/scala/akka/discovery/aggregate/AggregateServiceDiscovery.scala [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+import akka.event.Logging
+import akka.util.Helpers.Requiring
+import com.typesafe.config.Config
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+private final class AggregateServiceDiscoverySettings(config: Config) {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/aggregate/AggregateServiceDiscovery.scala]: is it visible for Java? then InternalApi markers

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+    override def apply(method: String): ServiceDiscovery = createServiceDiscovery(method)
+  }
+
+  private lazy val _defaultImplMethod =
+    system.settings.config.getString("akka.discovery.method") match {
+      case "<method>" ⇒
+        throw new IllegalArgumentException(
+          "No default service discovery implementation configured in " +
+            "`akka.discovery.method`. Make sure to configure this setting to your preferred implementation such as " +
+            "'akka-dns' in your application.conf (from the akka-discovery-dns module).")
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: akka-discovery module

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+object Discovery extends ExtensionId[Discovery] with ExtensionIdProvider {
+  override def apply(system: ActorSystem): Discovery = super.apply(system)
+
+  override def lookup: Discovery.type = Discovery
+
+  override def get(system: ActorSystem): Discovery = super.get(system)
+
+  override def createExtension(system: ExtendedActorSystem): Discovery = new Discovery()(system)
+
+  def checkClassPathForOldDiscovery(): Unit = {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: private

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+    try {
+      Class.forName("akka.discovery.SimpleServiceDiscovery")
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: all refection via `system.dynamicAccess`

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import akka.discovery._
+import akka.io.dns.DnsProtocol.{ Ip, Srv }
+import akka.io.dns.{ AAAARecord, ARecord, DnsProtocol, SRVRecord }
+
+import scala.collection.{ immutable ⇒ im }
+
+private object DnsServiceDiscovery {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: is it visible for Java? then InternalApi markers

File: akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+    Resolved(srvRequest, addresses)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+class DnsServiceDiscovery(system: ExtendedActorSystem) extends ServiceDiscovery {
```
Comment (patriknw) [inline akka-discovery/src/main/scala/akka/discovery/dns/DnsServiceDiscovery.scala]: `private[akka]`

**Review by patriknw — COMMENTED** [review]
some additional nits

File: akka-docs/src/main/paradox/discovery/index.md (hunk 15 above)
Comment (chbatey) [reply akka-docs/src/main/paradox/discovery/index.md]: Seems I just stalled here :)

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala (hunk 16 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: Yep, always under estimate the amount of publicity we get in Java

File: akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala (hunk 19 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/config/ConfigServiceDiscovery.scala]: Just keeping Raboof on his toes :)

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala (hunk 23 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: :+1: will use `getClassFor`

File: build.sbt [hunk 26]
```diff
@@ -474,6 +474,16 @@ lazy val actorTypedTests = akkaModule("akka-actor-typed-tests")
   .disablePlugins(MimaPlugin)
   .enablePlugins(NoPublish)
 
+lazy val discovery = akkaModule("akka-discovery")
```
Comment (patriknw) [inline build.sbt]: it should also be listed in `aggregatedProjects`

**Review by patriknw — APPROVED** [review]
LGTM after adding to aggregate in build

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+   * INTERNAL API
+   */
+  @InternalApi
+  private def createServiceDiscovery(method: String): ServiceDiscovery = {
+    val config = system.settings.config
+    val dynamic = system.dynamicAccess
+
+    def classNameFromConfig(path: String): String =
+      if (config.hasPath(path)) config.getString(path)
+      else "<nope>"
```
Comment (helena) [inline akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: Why do we do that, 'nope'?

File: akka-discovery/src/main/resources/reference.conf [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+  "/SD-DNS/async-dns" {
+    mailbox = "unbounded"
+    router = "round-robin-pool"
+    nr-of-instances = 1
+  }
+}
+
+akka.discovery {
+
+  # Users MUST configure this value to set the default discovery method.
```
Comment (helena) [inline akka-discovery/src/main/resources/reference.conf]:
```suggestion
  # Users MUST configure this value to set the default discovery mechanism.
```

**Comment by helena** [comment]
Helps organize and may affect [#25895](https://github.com/akka/akka/issues/25895) - in a good way.

File: akka-discovery/src/main/resources/reference.conf (hunk 28 above)
Comment (chbatey) [reply akka-discovery/src/main/resources/reference.conf]: we voted and decided to use `method` every where and remove `mechanism`

File: akka-discovery/src/main/scala/akka/discovery/Discovery.scala (hunk 27 above)
Comment (chbatey) [reply akka-discovery/src/main/scala/akka/discovery/Discovery.scala]: Ask he who should not be named! I am guessing it was to fail loading but i'll update this to explicitly fail

File: akka-discovery/src/main/resources/reference.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (raboof) [reply akka-discovery/src/main/resources/reference.conf]: Looks like this turned into `checkClassPathForOldDiscovery` below :+1:

File: akka-discovery/src/main/resources/reference.conf (hunk 28 above)
Comment (helena) [reply akka-discovery/src/main/resources/reference.conf]: 👍 k

**Review by helena — APPROVED** [review]
LGTM

File: akka-discovery/src/test/scala/akka/discovery/dns/DnsServiceDiscoverySpec.scala [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+      result.serviceName shouldEqual "cats.com"
+      result.addresses.toSet shouldEqual Set(
+        ResolvedTarget("kittens.com", Some(4), Some(InetAddress.getByName("127.0.0.2"))),
+        ResolvedTarget("kittens.com", Some(4), Some(InetAddress.getByName("127.0.0.3")))
+      )
+    }
+
+    // Naughty DNS server
+    "use SRV targetand port if no additional records" in {
```
Comment (raboof) [inline akka-discovery/src/test/scala/akka/discovery/dns/DnsServiceDiscoverySpec.scala]: missing space

File: akka-docs/src/main/paradox/discovery/index.md [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+
+Discovery used to be part of Akka Management but has become an Akka module as of `2.5.19` of Akka and version `0.21.0`
+of Akka Management. If you're also using Akka Management for other service discovery methods or bootstrap make
+sure you are using at least version `0.21.0`.
+
+## Dependency
+
+@@dependency[sbt,Gradle,Maven] {
+  group="com.typesafe.akka"
+  artifact="akka-discovery_2.12"
```
Comment (raboof) [inline akka-docs/src/main/paradox/discovery/index.md]: Use the `$scala.binary_version$` variable here

File: akka-docs/src/main/paradox/discovery/index.md [hunk 31]
```diff
@@ -0,0 +1,245 @@
...
+
+;; OPT PSEUDOSECTION:
+; EDNS: version: 0, flags:; udp: 4096
+; COOKIE: 5ab8dd4622e632f6190f54de5b28bb8fb1b930a5333c3862 (good)
+;; QUESTION SECTION:
+;service.tcp.akka.test.         IN      SRV
+
+;; ANSWER SECTION:
+service.tcp.akka.test.  86400   IN      SRV     10 60 5060 a-single.akka.test.
+service.tcp.akka.test.  86400   IN      SRV     10 40 5070 a-double.akka.test.
```
Comment (raboof) [inline akka-docs/src/main/paradox/discovery/index.md]: Do we want to add underscores here as well?

**Review by raboof — COMMENTED** [review]
Few, minor comments but looks almost good to go!

File: akka-docs/src/main/paradox/discovery/index.md (hunk 31 above)
Comment (chbatey) [reply akka-docs/src/main/paradox/discovery/index.md]: Yep

**Comment by raboof** [comment]
The failure seems Docker-related (`Failed to pull docker image [raboof/bind:9.11.3-20180713-nochown], is docker running?`)

**Comment by raboof** [comment]
PLS BUILD

**Comment by chbatey** [comment]
Thanks @raboof pushed a change for mima that should be sanity checked in case I just excluded all versions of everything from being checked for BC :)

**Review by patriknw — APPROVED** [review]
MiMa is looking good as far as I can tell, following same pattern as akka25

**Pull request merged**